src/psalm.php

```diff
@@ -56,810 +57,830 @@
 use const PHP_VERSION;
 use const STDERR;
 
-gc_collect_cycles();
-gc_disable();
-
 require_once __DIR__ . '/Psalm/Internal/ErrorHandler.php';
-ErrorHandler::install();
-
-require_once('command_functions.php');
+require_once __DIR__ . '/command_functions.php' ;
 require_once __DIR__ . '/Psalm/Internal/Composer.php';
+require_once __DIR__ . '/' . 'Psalm/Internal/IncludeCollector.php';
+require_once __DIR__ . '/' . 'Psalm/Internal/ExecutionEnvironment/BuildInfoCollector.php';
 
-
-$valid_short_options = [
-    'f:',
-    'm',
-    'h',
-    'v',
-    'c:',
-    'i',
-    'r:',
-];
-
-$valid_long_options = [
-    'clear-cache',
-    'clear-global-cache',
-    'config:',
-    'debug',
-    'debug-by-line',
-    'debug-performance',
-    'debug-emitted-issues',
-    'diff',
-    'disable-extension:',
-    'find-dead-code::',
-    'find-unused-code::',
-    'find-unused-variables',
-    'find-references-to:',
-    'help',
-    'ignore-baseline',
-    'init',
-    'memory-limit:',
-    'monochrome',
-    'no-diff',
-    'no-cache',
-    'no-reflection-cache',
-    'no-file-cache',
-    'output-format:',
-    'plugin:',
-    'report:',
-    'report-show-info:',
-    'root:',
-    'set-baseline:',
-    'show-info:',
-    'show-snippet:',
-    'stats',
-    'threads:',
-    'update-baseline',
-    'use-baseline:',
-    'use-ini-defaults',
-    'version',
-    'php-version:',
-    'generate-json-map:',
-    'generate-stubs:',
-    'alter',
-    'language-server',
-    'refactor',
-    'shepherd::',
-    'no-progress',
-    'long-progress',
-    'no-suggestions',
-    'include-php-versions', // used for baseline
-    'pretty-print', // used for JSON reports
-    'track-tainted-input',
-    'taint-analysis',
-    'security-analysis',
-    'dump-taint-graph:',
-    'find-unused-psalm-suppress',
-    'error-level:',
-];
-
-$args = array_slice($argv, 1);
-
-// get options from command line
-$options = getopt(implode('', $valid_short_options), $valid_long_options);
-
-if (isset($options['alter'])) {
-    require_once __DIR__ . '/psalter.php';
-    exit;
-}
-
-if (isset($options['language-server'])) {
-    require_once __DIR__ . '/psalm-language-server.php';
-    exit;
-}
-
-if (isset($options['refactor'])) {
-    require_once __DIR__ . '/psalm-refactor.php';
-    exit;
-}
-
-array_map(
+(
     /**
-     * @param string $arg
+     * @param array<int,string> $argv
+     *
+     * Note: the $argv parameter should not be renamed as it's used in the files this function includes via require_once
      */
-    function ($arg) use ($valid_long_options, $valid_short_options): void {
-        if (substr($arg, 0, 2) === '--' && $arg !== '--') {
-            $arg_name = preg_replace('/=.*$/', '', substr($arg, 2));
+    function (array $argv): void {
```

Comment:
We should probably move those closures into methods proper, e.g. `Psalm\Internal\Cli\Psalter::run()`